File: solutions/205-isomorphic-strings/solution.go
package main

import "fmt"

func isIsomorphic(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}

	sToT := make(map[byte]byte)
	tToS := make(map[byte]byte)

	for i := 0; i < len(s); i++ {
		a, b := s[i], t[i]
		if mapped, ok := sToT[a]; ok {
			if mapped != b {
				return false
			}
		} else {
			if _, ok := tToS[b]; ok {
				return false
			}
			sToT[a] = b
			tToS[b] = a
		}
	}
	return true
}

func main() {
	tests := []struct {
		s, t     string
		expected bool
	}{
		{"egg", "add", true},
		{"f11", "b23", false},
		{"paper", "title", true},
		{"badc", "baba", false},
	}

	for _, tc := range tests {
		result := isIsomorphic(tc.s, tc.t)
		status := "PASS"
		if result != tc.expected {
			status = "FAIL"
		}
		fmt.Printf("%s | s=%q, t=%q => %v (expected %v)\n", status, tc.s, tc.t, result, tc.expected)
	}
}

File: solutions/205-isomorphic-strings/solution.py
class Solution:
    def is_isomorphic(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False

        s_to_t: dict[str, str] = {}
        t_to_s: dict[str, str] = {}

        for a, b in zip(s, t):
            if a in s_to_t:
                if s_to_t[a] != b:
                    return False
            else:
                if b in t_to_s:
                    return False
                s_to_t[a] = b
                t_to_s[b] = a

        return True


if __name__ == "__main__":
    tests = [
        ("egg", "add", True),
        ("f11", "b23", False),
        ("paper", "title", True),
        ("badc", "baba", False),
    ]
    sol = Solution()
    for s, t, expected in tests:
        result = sol.is_isomorphic(s, t)
        status = "PASS" if result == expected else "FAIL"
        print(f"{status} | s={s!r}, t={t!r} => {result} (expected {expected})")
